src/dynamic_programming/knapsack.rs

```diff
@@ -1,148 +1,178 @@
 //! Solves the knapsack problem
-use std::cmp::max;
+use std::cmp::Ordering;
 
-/// knapsack_table(w, weights, values) returns the knapsack table (`n`, `m`) with maximum values, where `n` is number of items
+/// Solves the knapsack problem and returns the optimal profit, total weight, and indices of items included.
 ///
 /// # Arguments:
-///   * `w` - knapsack capacity
-///   * `weights` - set of weights for each item
-///   * `values` - set of values for each item
-fn knapsack_table(w: &usize, weights: &[usize], values: &[usize]) -> Vec<Vec<usize>> {
-    // Initialize `n` - number of items
-    let n: usize = weights.len();
-    // Initialize `m`
-    // m[i, w] - the maximum value that can be attained with weight less that or equal to `w` using items up to `i`
-    let mut m: Vec<Vec<usize>> = vec![vec![0; w + 1]; n + 1];
+///   * `capacity` - knapsack capacity
+///   * `item_weights` - weights of each item
+///   * `item_values` - values of each item
+///
+/// # Returns
+///   A tuple containing:
+///   - optimal profit
+///   - total weight of the optimal knapsack
+///   - indices of items included (from 1 to `num_items`)
+///
+/// # Complexity
+///   - Time complexity: O(num_items * capacity)
+///   - Space complexity: O(num_items * capacity)
+///
+/// where `num_items` is the number of items and `capacity` is the knapsack capacity.
+pub fn knapsack(
+    capacity: usize,
+    item_weights: Vec<usize>,
+    item_values: Vec<usize>,
+) -> (usize, usize, Vec<usize>) {
+    assert_eq!(item_weights.len(), item_values.len(), "Number of items in the list of weights doesn't match the number of items in the list of values!");
```

Comment:
Creating a struct named `Item` as follow: 
```rs
/// Represents an item with a weight and a value.
#[derive(Debug, PartialEq, Eq)]
pub struct Item {
    weight: usize,
    value: usize,
}
```
and a `KnapsackSolution` struct as follow:
```rs
/// Represents the solution to the knapsack problem.
#[derive(Debug, PartialEq, Eq)]
pub struct KnapsackSolution {
    /// The optimal profit obtained.
    optimal_profit: usize,
    /// The total weight of items included in the solution.
    total_weight: usize,
    /// The indices of items included in the solution. Indices might not be unique.
    item_indices: Vec<usize>,
}
```
help us prevent the exceptions like `InvalidInput` (miss matched length) because of we pass `Vec<Item>` into the `knapsack` function, each `Item` always has a weight and a value. And the cases where you suggested just work correctly as expected.

Thus, the knapsack function don't need to return an `Error`. It's make the function signature more clear and compact.
```rs
pub fn knapsack(capacity: usize, items: Vec<Item>) -> KnapsackSolution {
```
I really appreciate your practical suggestions (the last option). It's help the implementation more robust.